jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/deploy/DeployServerlessApplicationDialog.kt

```diff
@@ -0,0 +1,265 @@
+// Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package software.aws.toolkits.jetbrains.services.lambda.deploy
+
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.ui.DialogWrapper
+import com.intellij.openapi.ui.Messages
+import com.intellij.openapi.ui.ValidationInfo
+import software.amazon.awssdk.services.cloudformation.CloudFormationClient
+import software.amazon.awssdk.services.s3.S3Client
+import software.aws.toolkits.core.s3.regionForBucket
+import software.aws.toolkits.jetbrains.core.awsClient
+import software.aws.toolkits.jetbrains.core.region.AwsRegionProvider
+import software.aws.toolkits.jetbrains.services.cloudformation.Parameter
+import software.aws.toolkits.jetbrains.services.s3.CreateS3BucketDialog
+import software.aws.toolkits.jetbrains.utils.ui.addAndSelectValue
+import software.aws.toolkits.jetbrains.utils.ui.populateValues
+import software.aws.toolkits.jetbrains.utils.ui.selected
+import software.aws.toolkits.resources.message
+import java.awt.Dimension
+import java.awt.event.ActionEvent
+import javax.swing.Action
+import javax.swing.JComponent
+
+class DeployServerlessApplicationDialog(
+    private val project: Project,
+    private val parameters: Sequence<Parameter>
+) : DialogWrapper(project) {
+
+    private val view = DeployServerlessApplicationForm()
+    private val validator = DeploySamApplicationValidator()
+    private val s3Client: S3Client = project.awsClient()
+
+    private val regionProvider = AwsRegionProvider.getInstance()
+
+    private val deployAction = DeployServerlessApplicationAction()
+
+    init {
+        super.init()
+
+        title = message("serverless.application.deploy.title")
+
+        view.content.preferredSize = Dimension(800, 600)
+
+        view.withTemplateParameters(parameters.toList())
+
+        view.region.setRegions(regionProvider.regions().values.toMutableList())
+        view.btnCreateS3Bucket.isEnabled = view.region.selectedRegion != null
+
+        view.s3Bucket.populateValues {
+            emptyList()
+        }
+
+        view.stacks.addItem(getStackPlaceholderSelectRegion())
+
+        // S3 selector only shows buckets for region of interest
+        view.region.addActionListener {
+            view.btnCreateS3Bucket.isEnabled = view.region.selectedRegion != null
+            val activeRegionId = view.region.selectedRegion?.id
+
+            view.s3Bucket.populateValues {
+                if (!activeRegionId.isNullOrEmpty()) {
+                    s3Client.listBuckets().buckets()
```

Comment:
I feel like we use this in a few places, might be time for a `S3Client.listBuckets(region: Region)` extension function.